File: agent-workspace-package/templates/skills/workspace-init/scripts/link_codebases.py
#!/usr/bin/env python3
"""Link sibling codebases into codebase-symlinks/ and refresh the codebase map."""

from __future__ import annotations

import argparse
import json
import os
import re
import sys
from dataclasses import dataclass
from pathlib import Path

CELL_TEXT_LIMIT = 120


@dataclass(frozen=True)
class Candidate:
    name: str
    path: Path
    summary: str
    source_note: str


@dataclass(frozen=True)
class LinkedCodebase:
    name: str
    source: Path
    symlink: Path
    purpose: str
    source_note: str


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(
        description=(
            "Interactively link sibling codebases one directory above the workspace "
            "into codebase-symlinks/ and regenerate the codebase map."
        )
    )
    parser.add_argument(
        "workspace",
        nargs="?",
        default=".",
        help="Workspace root directory. Defaults to the current directory.",
    )
    return parser.parse_args()


def normalize_text(value: str, limit: int | None = None) -> str:
    cleaned = " ".join(value.split()).replace("|", "&#124;")
    if limit is not None and len(cleaned) > limit:
        cleaned = cleaned[: max(0, limit - 3)].rstrip() + "..."
    return cleaned


def read_text(path: Path) -> str:
    try:
        return path.read_text(encoding="utf-8", errors="ignore")
    except OSError:
        return ""


def extract_readme_summary(path: Path) -> str:
    text = read_text(path)
    if not text:
        return ""

    lines = text.splitlines()
    paragraph: list[str] = []
    saw_heading = False

    for raw_line in lines:
        line = raw_line.strip()
        if not line:
            if paragraph:
                break
            continue
        if line.startswith("#"):
            saw_heading = True
            continue
        if saw_heading:
            paragraph.append(line)

    if paragraph:
        return normalize_text(" ".join(paragraph), CELL_TEXT_LIMIT)

    headings = [line.lstrip("#").strip() for line in lines if line.strip().startswith("#")]
    for heading in headings:
        if heading:
            return normalize_text(heading, CELL_TEXT_LIMIT)

    nonempty = [line.strip() for line in lines if line.strip()]
    if nonempty:
        return normalize_text(nonempty[0], CELL_TEXT_LIMIT)

    return ""


def extract_json_description(path: Path) -> str:
    try:
        data = json.loads(read_text(path))
    except json.JSONDecodeError:
        return ""

    for key in ("description", "name"):
        value = data.get(key)
        if isinstance(value, str) and value.strip():
            return normalize_text(value.strip(), CELL_TEXT_LIMIT)

    return ""


def extract_toml_field(path: Path, field: str) -> str:
    pattern = re.compile(rf'^\s*{re.escape(field)}\s*=\s*["\'](.+?)["\']\s*$')
    for raw_line in read_text(path).splitlines():
        match = pattern.match(raw_line)
        if match:
            return normalize_text(match.group(1).strip(), CELL_TEXT_LIMIT)
    return ""


def summarize_repository(path: Path) -> tuple[str, str]:
    if not path.exists():
        return "Target missing", "missing target"

    readme = path / "README.md"
    if readme.is_file():
        summary = extract_readme_summary(readme)
        if summary:
            return summary, "README.md"

    package_json = path / "package.json"
    if package_json.is_file():
        summary = extract_json_description(package_json)
        if summary:
            return summary, "package.json"

    pyproject = path / "pyproject.toml"
    if pyproject.is_file():
        summary = extract_toml_field(pyproject, "description") or extract_toml_field(
            pyproject, "name"
        )
        if summary:
            return summary, "pyproject.toml"

    cargo = path / "Cargo.toml"
    if cargo.is_file():
        summary = extract_toml_field(cargo, "description") or extract_toml_field(
            cargo, "name"
        )
        if summary:
            return summary, "Cargo.toml"

    return normalize_text(path.name, CELL_TEXT_LIMIT), "directory name"


def discover_sibling_codebases(workspace_root: Path) -> list[Candidate]:
    source_parent = workspace_root.parent
    candidates: list[Candidate] = []
    workspace_resolved = workspace_root.resolve()

    for path in sorted(source_parent.iterdir(), key=lambda item: item.name.lower()):
        if not path.is_dir():
            continue

        try:
            if path.resolve() == workspace_resolved:
                continue
        except OSError:
            continue

        summary, source_note = summarize_repository(path)
        candidates.append(
            Candidate(
                name=path.name,
                path=path.resolve(),
                summary=summary,
                source_note=source_note,
            )
        )

    return candidates


def print_candidates(candidates: list[Candidate], source_parent: Path) -> None:
    print(f"Scanning sibling directories in: {source_parent}")
    print("Available codebases:")
    for index, candidate in enumerate(candidates, start=1):
        print(f"  {index}. {candidate.name}")
        print(f"     Purpose: {candidate.summary}")
        print(f"     Source: {candidate.source_note}")


def prompt_selection(candidates: list[Candidate]) -> list[Candidate]:
    while True:
        raw = input(
            "Select codebases to symlink (all, comma-separated numbers, or q to quit): "
        ).strip()

        if not raw:
            print("Please choose at least one codebase, or type q to quit.")
            continue

        lowered = raw.lower()
        if lowered in {"q", "quit", "exit"}:
            return []

        if lowered in {"a", "all", "*"}:
            return candidates

        chosen_indexes: list[int] = []
        tokens = [token for token in re.split(r"[,\s]+", raw) if token]
        valid = True
        for token in tokens:
            if not token.isdigit():
                valid = False
                break
            index = int(token)
            if index < 1 or index > len(candidates):
                valid = False
                break
            if index not in chosen_indexes:
                chosen_indexes.append(index)

        if not valid or not chosen_indexes:
            print("Invalid selection. Use numbers, a comma-separated list, or all.")
            continue

        return [candidates[index - 1] for index in chosen_indexes]


def link_codebase(source: Path, destination: Path) -> str:
    destination.parent.mkdir(parents=True, exist_ok=True)

    if destination.exists() or destination.is_symlink():
        if destination.is_symlink():
            try:
                current_target = destination.resolve(strict=False)
            except OSError:
                current_target = None
            if current_target == source.resolve():
                print(f"Preserved existing link: {destination} -> {source}")
                return "preserved"

        print(f"Skipping existing path: {destination}")
        return "skipped"

    relative_target = os.path.relpath(source, destination.parent)
    destination.symlink_to(relative_target, target_is_directory=True)
    print(f"Created symlink: {destination} -> {relative_target}")
    return "created"


def collect_linked_codebases(destination_root: Path) -> list[LinkedCodebase]:
    if not destination_root.exists():
        return []

    linked: list[LinkedCodebase] = []
    for path in sorted(destination_root.iterdir(), key=lambda item: item.name.lower()):
        if not path.is_symlink():
            print(f"Warning: skipping non-symlink entry in {destination_root}: {path.name}")
            continue

        source = path.resolve(strict=False)
        purpose, source_note = summarize_repository(source)
        linked.append(
            LinkedCodebase(
                name=path.name,
                source=source,
                symlink=path,
                purpose=purpose,
                source_note=source_note,
            )
        )

    return linked


def render_codebase_map(
    entries: list[LinkedCodebase], workspace_root: Path, destination_root: Path
) -> str:
    symlink_root = destination_root.relative_to(workspace_root)
    lines = [
        "# Codebase Map",
        "",
        "This note records sibling repositories linked into `codebase-symlinks/` so agents can identify the right codebase before traversing back to the parent project.",
        "",
        f"- Symlink root: `{normalize_text(str(symlink_root), CELL_TEXT_LIMIT)}`",
        "",
        "| Codebase | Symlink | Purpose | Source note |",
        "| --- | --- | --- | --- |",
    ]

    if not entries:
        lines.append("| _None_ | _None_ | No linked codebases recorded yet. | - |")
        lines.extend(
            [
                "",
                "Regenerate this note after linking sibling codebases through the workspace initializer.",
            ]
        )
        return "\n".join(lines) + "\n"

    for entry in entries:
        lines.append(
            "| "
            + " | ".join(
                [
                    f"`{normalize_text(entry.name, CELL_TEXT_LIMIT)}`",
                    f"`{normalize_text(str(entry.symlink.relative_to(workspace_root)), CELL_TEXT_LIMIT)}`",
                    normalize_text(entry.purpose, CELL_TEXT_LIMIT),
                    normalize_text(entry.source_note, CELL_TEXT_LIMIT),
                ]
            )
            + " |"
        )

    lines.extend(
        [
            "",
            "Keep the purpose line short and specific when the linked repos change.",
        ]
    )
    return "\n".join(lines) + "\n"


def refresh_codebase_map(workspace_root: Path, destination_root: Path) -> Path:
    map_path = workspace_root / "knowledgebase" / "context-history" / "codebase-map.md"
    map_path.parent.mkdir(parents=True, exist_ok=True)

    entries = collect_linked_codebases(destination_root)
    map_path.write_text(
        render_codebase_map(entries, workspace_root, destination_root), encoding="utf-8"
    )
    return map_path


def main() -> int:
    args = parse_args()
    workspace_root = Path(args.workspace).expanduser().resolve()

    if not workspace_root.is_dir():
        print(f"Error: workspace directory not found: {workspace_root}", file=sys.stderr)
        return 1

    source_parent = workspace_root.parent
    destination_root = workspace_root / "codebase-symlinks"

    candidates = discover_sibling_codebases(workspace_root)
    if not candidates:
        print(f"No sibling directories found in {source_parent}.")
        return 0

    print(f"Workspace root: {workspace_root}")
    print(f"Symlink destination: {destination_root}")
    print_candidates(candidates, source_parent)

    try:
        selected = prompt_selection(candidates)
    except KeyboardInterrupt:
        print("\nSelection canceled.")
        return 130
    except EOFError:
        print("\nSelection canceled.")
        return 130

    if not selected:
        print("No codebases selected. Nothing linked.")
        if destination_root.exists():
            map_path = refresh_codebase_map(workspace_root, destination_root)
            print(f"Refreshed codebase map: {map_path}")
        return 0

    created = 0
    preserved = 0
    skipped = 0

    for candidate in selected:
        status = link_codebase(candidate.path, destination_root / candidate.name)
        if status == "created":
            created += 1
        elif status == "preserved":
            preserved += 1
        else:
            skipped += 1

    map_path = refresh_codebase_map(workspace_root, destination_root)

    print(
        "Linking complete: "
        f"{created} created, {preserved} preserved, {skipped} skipped."
    )
    print(f"Updated codebase map: {map_path}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: agent-workspace-package/src/core/link-codebases.js
import { readdirSync, readFileSync, writeFileSync, mkdirSync, existsSync, lstatSync, symlinkSync, realpathSync, statSync } from 'node:fs';
import { join, relative, resolve as resolvePath, basename } from 'node:path';
import { createInterface } from 'node:readline/promises';

const CELL_LIMIT = 120;

function normalizeText(val, limit) {
  const cleaned = val.replace(/\s+/g, ' ').replace(/\|/g, '&#124;').trim();
  if (limit && cleaned.length > limit) return cleaned.slice(0, limit - 3).trimEnd() + '...';
  return cleaned;
}

function readText(path) {
  try { return readFileSync(path, 'utf8'); } catch { return ''; }
}

function extractReadmeSummary(dir) {
  const text = readText(join(dir, 'README.md'));
  if (!text) return '';
  const lines = text.split('\n');
  const para = [];
  let sawHeading = false;
  for (const raw of lines) {
    const line = raw.trim();
    if (!line) { if (para.length) break; continue; }
    if (line.startsWith('#')) { sawHeading = true; continue; }
    if (sawHeading) para.push(line);
  }
  if (para.length) return normalizeText(para.join(' '), CELL_LIMIT);
  const headings = lines.filter(l => l.trim().startsWith('#')).map(l => l.replace(/^#+\s*/, '').trim());
  for (const h of headings) if (h) return normalizeText(h, CELL_LIMIT);
  const nonempty = lines.map(l => l.trim()).filter(Boolean);
  return nonempty[0] ? normalizeText(nonempty[0], CELL_LIMIT) : '';
}

function extractJsonDescription(dir) {
  try {
    const data = JSON.parse(readText(join(dir, 'package.json')));
    for (const key of ['description', 'name']) {
      if (typeof data[key] === 'string' && data[key].trim()) return normalizeText(data[key].trim(), CELL_LIMIT);
    }
  } catch { /* */ }
  return '';
}

function extractTomlField(dir, file, field) {
  const text = readText(join(dir, file));
  const pattern = new RegExp(`^\\s*${field}\\s*=\\s*["'](.+?)["']\\s*$`, 'm');
  const m = text.match(pattern);
  return m ? normalizeText(m[1].trim(), CELL_LIMIT) : '';
}

function summarize(dir) {
  return extractReadmeSummary(dir)
    || extractJsonDescription(dir)
    || extractTomlField(dir, 'pyproject.toml', 'description')
    || extractTomlField(dir, 'pyproject.toml', 'name')
    || extractTomlField(dir, 'Cargo.toml', 'description')
    || extractTomlField(dir, 'Cargo.toml', 'name')
    || normalizeText(basename(dir), CELL_LIMIT);
}

function discoverSiblings(workspaceRoot) {
  const parent = join(workspaceRoot, '..');
  const resolvedRoot = resolvePath(workspaceRoot);
  const results = [];
  try {
    for (const name of readdirSync(parent).sort((a, b) => a.toLowerCase().localeCompare(b.toLowerCase()))) {
      const full = resolvePath(join(parent, name));
      try {
        if (!statSync(full).isDirectory()) continue;
        if (full === resolvedRoot) continue;
        results.push({ name, path: full, summary: summarize(full) });
      } catch { /* */ }
    }
  } catch { /* */ }
  return results;
}

async function promptSelection(candidates) {
  const iface = createInterface({ input: process.stdin, output: process.stdout });
  try {
    console.log('\nAvailable sibling codebases:');
    candidates.forEach((c, i) => {
      console.log(`  ${i + 1}. ${c.name}`);
      console.log(`     ${c.summary}`);
    });
    const raw = (await iface.question('\nSelect (all, comma-separated numbers, or q to skip): ')).trim().toLowerCase();
    if (!raw || raw === 'q' || raw === 'quit') return [];
    if (raw === 'all' || raw === 'a' || raw === '*') return candidates;
    const parts = raw.split(/[\s,]+/).filter(Boolean);
    const selected = [];
    for (const p of parts) {
      const idx = parseInt(p, 10);
      if (!isNaN(idx) && idx >= 1 && idx <= candidates.length) selected.push(candidates[idx - 1]);
    }
    return [...new Map(selected.map(c => [c.name, c])).values()];
  } finally {
    iface.close();
  }
}

function linkCodebase(source, dest) {
  mkdirSync(join(dest, '..'), { recursive: true });
  let existsStat;
  try { existsStat = lstatSync(dest); } catch { existsStat = null; }

  if (existsStat) {
    if (existsStat.isSymbolicLink()) {
      try {
        if (realpathSync(dest) === source) return 'preserved';
      } catch { /* */ }
    }
    return 'skipped';
  }

  const rel = relative(join(dest, '..'), source);
  symlinkSync(rel, dest);
  return 'created';
}

function renderMap(entries, workspaceRoot, symlinkRoot) {
  const relRoot = relative(workspaceRoot, symlinkRoot);
  const lines = [
    '# Codebase Map',
    '',
    'This note records sibling repositories linked into `codebase-symlinks/` so agents can identify the right codebase before traversing back to the parent project.',
    '',
    `- Symlink root: \`${relRoot}\``,
    '',
    '| Codebase | Symlink | Purpose |',
    '| --- | --- | --- |',
  ];

  if (!entries.length) {
    lines.push('| _None_ | _None_ | No linked codebases recorded yet. |');
  } else {
    for (const e of entries) {
      const symRel = relative(workspaceRoot, e.symlink);
      lines.push(`| \`${e.name}\` | \`${symRel}\` | ${e.summary} |`);
    }
  }

  lines.push('', 'Keep the purpose line short and specific when the linked repos change.');
  return lines.join('\n') + '\n';
}

function refreshMap(workspaceRoot, symlinkRoot) {
  const mapPath = join(workspaceRoot, 'knowledgebase', 'context-history', 'codebase-map.md');
  mkdirSync(join(mapPath, '..'), { recursive: true });

  const entries = [];
  if (existsSync(symlinkRoot)) {
    for (const name of readdirSync(symlinkRoot).sort()) {
      const p = join(symlinkRoot, name);
      try {
        if (!lstatSync(p).isSymbolicLink()) continue;
        const source = resolvePath(p);
        entries.push({ name, symlink: p, summary: summarize(source) });
      } catch { /* */ }
    }
  }

  writeFileSync(mapPath, renderMap(entries, workspaceRoot, symlinkRoot), 'utf8');
  return mapPath;
}

export async function linkCodebases(workspaceRoot, { linkAll = false, dryRun = false } = {}) {
  const symlinkRoot = join(workspaceRoot, 'codebase-symlinks');
  const candidates = discoverSiblings(workspaceRoot);

  if (!candidates.length) {
    console.log('No sibling directories found.');
    return { created: 0, preserved: 0, skipped: 0 };
  }

  let selected;
  if (linkAll) {
    selected = candidates;
  } else {
    try {
      selected = await promptSelection(candidates);
    } catch {
      selected = [];
    }
  }

  if (!selected.length) {
    if (!dryRun) refreshMap(workspaceRoot, symlinkRoot);
    return { created: 0, preserved: 0, skipped: 0 };
  }

  const counts = { created: 0, preserved: 0, skipped: 0 };
  if (!dryRun) mkdirSync(symlinkRoot, { recursive: true });

  for (const c of selected) {
    const dest = join(symlinkRoot, c.name);
    const status = dryRun ? 'created' : linkCodebase(c.path, dest);
    counts[status] = (counts[status] ?? 0) + 1;
    console.log(`  ${status === 'created' ? '+' : status === 'preserved' ? '=' : '↷'} ${c.name}`);
  }

  if (!dryRun) {
    const mapPath = refreshMap(workspaceRoot, symlinkRoot);
    console.log(`  Updated codebase map: ${relative(workspaceRoot, mapPath)}`);
  }

  return counts;
}
